### DISCUSSION TITLE
simple audio volume controls

### DISCUSSION THREAD

**Discussion by marc-portier** [micro-nova/AmpliPi#203] [Q&A] [answered]
I (think) I get that controlling the AmpliPi is in first place to be done through the restAPI and a web-front-end that can be controlled from any device connected to the network. Allowing users to select sources to be played in which zones + control the volumes.

In some of my zones however it seems useful to have simple hardware-access to a basic volume control composed of

1. a on/off-like control for mute/unmute
2. a dimmer-like control for increase/decrease volume

a fancy 3rd option might be prev-next looping through a list of managed "favorite sources" 

I have read #77 but don't understand all of it, plus I think the additional display and controller is over the top for what I need.  And I don't seem to find another topic that better aligns to my needs?

My own thoughts for solving it go towards wiring the buttons from the zones to a central rasbpi (next to the AmpliPi?) that would translate contacts from those buttons into REST calls on the AmpliPi? But maybe there are other / simpler ways to address this?  Any thoughts?

Disclaimer: I know how to code, but hardware wiring has not been my thing - still the basics for these kind of simple buttons are probably nicely handled by basic "rasbpi IO examples" so probably within?  Anyway don't be shy on refering to entry level stuff or being overly explicit on that part, it will help me for sure.

**Comment by Lohrer** [comment, marked as answer]
It sounds like you're looking for either a very simple wall-plate device or something like a remote, is that right? In #77 the idea is an in-wall touchscreen that can do all the controls you mentioned plus show the current status. It communicates back to AmpliPi via the ESP32's WiFi or wired Ethernet connection. I think the hardware you want is basically that but remove the screen and replace with some buttons.
1. mute/unmute or play/pause can either be a single button that toggles the state, one button for each state, or a switch.
2. increase/decrease volume can easily be two buttons, or a paddle like on smart dimmer switches (which is just two buttons).
3. There isn't any concept of "favorite sources" (yet?), but there are [presets](https://api.amplipi.com/#post-/api/preset) that could help with this. A button could cycle through presets, or there couple be several buttons available to be assigned to presets.

There are a lot of possibilities here for controls. It might even make sense to just make hardware with several generic buttons and assign them functions as needed. That would allow changing things later, or even adding control for other devices.

For connecting back to AmpliPi, you could wire all the buttons back to a central RPi, but if you don't already have wiring it would probably be easier using WiFi with something like the ESP32. Then you just need power. It might even be possible to battery-power it but I'm not sure of the power draw of the ESP32.

If you wire buttons all the way back to a central location you definitely could use a second Raspberry Pi to read them all and forward the appropriate message on to the AmpliPi API. If you wire back to the AmpliPi, you could even just add some USB GPIO device (an Arduino?) and run your button-reading code on the AmpliPi itself. I doubt many people want to add all that wiring though, which is why #77 is using WiFi, or PoE which may already be ran.

This is a very open-ended idea, so hopefully I've helped a bit here but if not feel free to ask more specific questions. We'd love to see some hardware control for AmpliPi! There may even be some off-the-shelf remote that could be re-programmed to send AmpliPi messages.

**Reply by marc-portier** [reply]
Wow, this is an interesting new world opening up for me :) 

In my case I have the luxury/burden of a naked/stripped-down house that requires new wires for everything anyway,
Loads of options it seems.

Trying to rehash what I understood (combining some quick google searches and maiden-read-ups)

1. USB GPIO device(s)
- sounds like a mini controller that bridges from IO pins to USB input 
  - can be programmed to read physical IO from various buttons
  - convert that into characters / messages sent to the AmpliPi (as if it would be some external usb-keyboard)
- assumes as wiring:
  - all buttons from all zones (say 6 x 4 = mute-toggle, up-down, next-preset) down to the usb gpio - most likely centrally, close to the AmpliPi
  - usb connection from gpio to amplipi
- assumes as software:
  - straightforward mapping of the IO buttons/contacts to characters (messages)  = code on the controller
  - the ampli-pi running an extra module reading input from the usb and calling API ?


2. ESP32
- sounds like a more advanced self-sufficient beast bridging from IO pins to network calls 
  - can be programmed to read physical IO from various buttons
  - convert that into full fledge REST API calls sent over the network
  - or as suggested in #77 support a full two way touch screen 
- assumes as wiring:
  - IO inputs from buttons to ESP32 - more likely close to the buttons, close to the zones
  - either only power + use available wifi / either UTP carrying POE to the ESP32
- assumes software:
  - only on the ESP32 since it can use the network to call to the REST API of AmpliPi



What about trustworthiness and guaranteed lifetime of these components?
Will these easily keep on working 2-5-10-20 years?
How easy will the replacement or reprogramming turn out to be?


Any further thoughts welcome. But already grateful for the provided pointers.

**Reply by Lohrer** [reply]
### USB
Sounds like you have a good idea of what I was going for here. To answer your question, yes the AmpliPi controller (a Raspberry Pi) would run some extra module that reads input from the USB device and send the appropriate AmpliPi API message when a button is pressed.

### WiFi/Ethernet
Sounds like you get there idea here now too! You can totally piggy-back off the work in that thread, and we've played with this solution a bit at MicroNova too.

Whatever you choose, the general idea is going to be a something that reads a button state via GPIO and sends an Ethernet messages to AmpliPi's API. Whether it's done on the AmpliPi's controller itself, a second Raspberry Pi, an ESP32 or something else entirely is really just an implementation choice.

I would like to add though that the downside to running wires to all your buttons ahead of time is you can't add more buttons or change up the config later. A suggestion would be to run CAT 5/6 cabling even if you only plan on using it for direct button reading for now. Then you have more options for upgrading to Ethernet later. HOWEVER, I'm not experienced with extending GPIO that far with any wiring, let alone twisted pairs. Maybe someone else can chime in on that.

**Reply by marc-portier** [reply]
thx, more smart ideas to consider when making the upcoming wire-decision!
